ui/objects/semantic-ui/js/objects/editform/components/VocabularyTreeSelectField.jsx

```diff
@@ -0,0 +1,428 @@
+      setSelectedState((prevState) => {
+        const newState = [...prevState];
+        newState.splice(existingIndex, 1);
+        return newState;
+      });
+    } else {
+      if (multiple && selectedState.length != 0) {
+        setSelectedState((prevState) => [...prevState, option]);
+      } else {
+        setSelectedState([option]);
+      }
+    }
+  };
+
+
+  const handleSubmit = () => {
+    let prepSelect;
+    if (multiple) {
+      prepSelect = selectedState.map((i) => serializeVocabularyItem(i.value));
+      const existingValues = getIn(formik.values, fieldPath, []);
+      prepSelect = [...existingValues, ...prepSelect];
+    } else {
+      prepSelect = selectedState.map((i) =>
+        serializeVocabularyItem(i.value)
+      )[0];
+    }
+
+    const updatedValues = setIn(formik.values, fieldPath, prepSelect);
+
+    formik.setValues(updatedValues);
+    setOpenState(false);
+    setSelectedState([]);
+    setParentsState([]);
+  };
+
+  const handleKey = (e, option, index) => {
+    e.preventDefault();
+    let newIndex = 0;
+
+    index = keybState.length - 1 > index ? keybState.length - 1 : index;
+    let data = hierarchicalData[index];
+
+    const moveKey = (index, newIndex, back = false) => {
+    
+      setKeybState((prev) => {
+        const newState = [...prev];
+        back ? newState.splice(index, 1) : (newState[index] = newIndex);
+        return newState;
+      });
+    };
+    if (e.key === "ArrowUp") {
+      newIndex = keybState[index] - 1;
+      if (newIndex >= 0) {
+        updateHierarchy(data[newIndex].value, index)();
+        moveKey(index, newIndex, false);
+      }
+    } else if (e.key === "ArrowDown") {
+      newIndex = keybState[index] + 1;
+
+      if (newIndex < data.length) {
+        updateHierarchy(data[newIndex].value, index)();
+        moveKey(index, newIndex, false);
+      }
+    } else if (e.key === "ArrowLeft") {
+      if (index > 0) {
+        setParentsState((prev) => {
+          const newState = [...prev];
+          newState.splice(index, 1);
+          return newState;
+        });
+        moveKey(index, null, true);
+      }
+    } else if (e.key === "ArrowRight") {
+      if (index < amount - 1) {
+        const nextColumnOptions = hierarchicalData[index + 1];
+        if (nextColumnOptions) {
+          const nextColumnIndex = nextColumnOptions.findIndex(
+            (o) => o.hierarchy.ancestors[0] === parentsState[index]
+          );
+          if (nextColumnIndex !== -1) {
+            newIndex = nextColumnIndex;
+            updateHierarchy(
+              nextColumnOptions[nextColumnIndex].value,
+              index + 1
+            )();
+            moveKey(index + 1, newIndex, false);
+          }
+        }
+      }
+    } else if (e.key === "Enter") {
+      handleSelect(data[keybState[index]], index, e);
+    } else if ((e.key = " ")) {
+      handleSelect(data[keybState[index]], index, e);
+    }
+  };
+
+  const renderColumns = (index) => {
```

Comment:
This function should be just `renderColumn` or should be extracted to separate component `HierarchyColumn` that accepts `column` items and hierarchy `level` prop and renders a Grid.Columns out of it.